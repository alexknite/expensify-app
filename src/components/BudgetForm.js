import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { DateRangePicker } from 'react-dates';
import { startAddCategory } from '../actions/categories';

export class BudgetForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      category: props.budget ? props.budget.category : '',
      amount: props.budget ? (props.budget.amount / 100).toString() : '',
      startDate: props.budget ? moment(props.budget.startDate) : moment(),
      endDate: props.budget ? moment(props.budget.endDate) : moment().add(7, 'days'),
      note: props.budget ? props.budget.note : '',
      calendarFocused: null
    };
  };
  onCategoryChange = (e) => {
    const category = e.target.value;
    this.setState(() => ({ category }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.setState(() => ({ startDate, endDate }));
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onSubmit = async (e) => {
    e.preventDefault();
    const cleansedCategories = this.props.categories.map(({ name }) => name
      .replace(/[^A-Z0-9]+/ig, '')
      .toLowerCase());
    const cleansedCategory = this.state.category
      .replace(/[^A-Z0-9]+/ig, '')
      .toLowerCase();
    const categoryIndex = cleansedCategories.indexOf(cleansedCategory);

    if (categoryIndex > -1) {
      await this.setState(() => ({ category: this.props.categories[categoryIndex].name }));
    } else {
      this.props.startAddCategory({ name: this.state.category });
    }

    if (!this.state.amount || (!this.state.startDate && !this.state.endDate) ) {
      this.setState(() => ({ error: 'Please provide the start date, end date and the amount' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        category: this.state.category,
        amount: parseFloat(this.state.amount, 10) * 100,
        startDate: this.state.startDate.valueOf(),
        endDate: this.state.endDate.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Category"
          autoFocus
          value={this.state.category}
          onChange={(this.onCategoryChange)}
        />
        <input
          className="text-input"
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <DateRangePicker
          startDate={this.state.startDate}
          endDate={this.state.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
        />
        <textarea
          className="textarea"
          placeholder="Add a note for your budget (optional)"
          value={this.state.note}
          onChange={(this.onNoteChange)}
        >
        </textarea>
        <div>
          <button className="button">Save Budget</button>
        </div>
      </form>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    categories: state.categories
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    startAddCategory: ({ name }) => dispatch(startAddCategory({ name }))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(BudgetForm);
