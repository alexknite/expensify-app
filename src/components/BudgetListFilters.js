import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setBudgetsTextFilter, sortBudgetsByDate, sortBudgetsByAmount, setBudgetsStartDate, setBudgetsEndDate, setBudgetsCategoryFilter } from '../actions/budgetsFilters';

export class BudgetListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setBudgetsStartDate(startDate);
    this.props.setBudgetsEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortBudgetsByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortBudgetsByAmount();
    }
  };
  onCategoryChange = (e) => {
    this.props.setBudgetsCategoryFilter(e.target.value);
  };
  render() {
    return (
      <div className="content-container content-container__filters">
        <div className="input-group">
          <div className="input-group__item">
            {
              this.props.categories.length === 0
              ? <span></span>
              : (
                <select
                  className="select"
                  value={this.props.budgetsFilters.category}
                  onChange={this.onCategoryChange}
                >
                  <option
                    disabled
                    value
                  >
                    -- Select a category --
                  </option>
                {
                  this.props.categories.map(({ id, name }) => {
                    return <option
                      key={id}
                      value={id}
                    >
                      {name}
                    </option>
                  })
                }
                <option value="">None</option>
                </select>
              )
            }
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.budgetsFilters.sortBy}
              onChange={this.onSortChange}
            >
              <option
                value="date"
              >
                Date
              </option>
              <option
                value="amount"
              >
                Amount
              </option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.budgetsFilters.startDate}
              endDate={this.props.budgetsFilters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  budgetsFilters: state.budgetsFilters,
  categories: state.categories
});

const mapDispatchToProps = (dispatch) => ({
  setBudgetsTextFilter: (text) => dispatch(setBudgetsTextFilter(text)),
  sortBudgetsByDate: () => dispatch(sortBudgetsByDate()),
  sortBudgetsByAmount: () => dispatch(sortBudgetsByAmount()),
  setBudgetsStartDate: (startDate) => dispatch(setBudgetsStartDate(startDate)),
  setBudgetsEndDate: (endDate) => dispatch(setBudgetsEndDate(endDate)),
  setBudgetsCategoryFilter: (category) => dispatch(setBudgetsCategoryFilter(category))
});

export default connect(mapStateToProps, mapDispatchToProps)(BudgetListFilters);
