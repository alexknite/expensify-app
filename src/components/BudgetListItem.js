import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { selectBudget } from '../actions/budget';

const BudgetListItem = ({ id, amount, startDate, endDate, note, selectBudget, category }) => {
  const onSelect = () => {
    selectBudget({
      id,
      amount,
      startDate,
      endDate,
      note,
      category
    });
  };
  return (
    <div className="list-item">
      <div>
        <h2 className="list-item__sub-title">{category}</h2>
        <h3 className="list-item__title">{numeral(amount / 100).format(',0.00')}</h3>
        <p className="list-item__sub-title">{moment(startDate).format('MMMM Do, YYYY')} - {moment(endDate).format('MMMM Do, YYYY')}</p>
        <h3 className="list-item__sub-title">{note}</h3>
      </div>
      <div className="list-item__actions">
        <button className="button button--link" onClick={onSelect}>
          <img src="/images/eye.svg" alt="Select"/>
        </button>
        <Link to={`edit/budget/${id}`}>
          <button className="button button--link">
            <img src="/images/edit.svg" alt="Edit"/>
          </button>
        </Link>
      </div>
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    selectBudget: (budget) => dispatch(selectBudget(budget))
  };
};

export default connect(undefined, mapDispatchToProps)(BudgetListItem);
