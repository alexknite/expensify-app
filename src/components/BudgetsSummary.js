import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectBudget from '../selectors/budgets-diff';

export const BudgetsSummary = ({ moneySpent, moneyAlotted, moneyLeft, category }) => {
  const formattedMoneySpent = numeral(moneySpent / 100).format('$0,0.00');
  const formattedMoneyAlotted = numeral(moneyAlotted / 100).format('$0,0.00');
  const formattedMoneyLeft = numeral(moneyLeft / 100).format('$0,0.00');

  return (
    <div className="page-header">
      <div className="content-container">
        {
          moneyAlotted && category ? (
            <div>
              <h1 className="page-header__title">You have spent <span>{formattedMoneySpent}</span> out of <span>{formattedMoneyAlotted}</span> for <span>{category}</span></h1>
              {
                moneyLeft <= 0 ? (
                  <h1 className="page-header__title">You have spent all your money for <span>{category}</span>!</h1>
                ) : (
                  <h1 className="page-header__title">You may spend <span>{formattedMoneyLeft}</span> more</h1>
                )
              }
            </div>
          ) : (
            <span></span>
          )
        }
        <div className="page-header__actions">
          <Link className="button" to="/create/budget">Add Budget</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = ({ budget, expenses }) => {
  const moneySpent = selectBudget(expenses, budget.category, budget.amount);
  const moneyAlotted = budget.amount;
  const moneyLeft = moneyAlotted - moneySpent;

  return {
    moneySpent,
    moneyAlotted,
    moneyLeft,
    category: budget.category
  };
};

export default connect(mapStateToProps)(BudgetsSummary);
