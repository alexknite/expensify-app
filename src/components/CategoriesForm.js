import React from 'react';

export class CategoriesForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: props.category ? props.category.name : ''
    };
    this.onTextChange = this.onTextChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  };
  onTextChange(e) {
    const name = e.target.value;
    this.setState(() => ({ name }));
  };
  onSubmit(e) {
    e.preventDefault();

    if (this.state.name === '') {
      this.setState(() => ({ error: 'Category names cannot be blank' }));
    } else {
      this.props.onSubmit({
        name: this.state.name
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Category"
          autoFocus
          value={this.state.name}
          onChange={(this.onTextChange)}
        />
        <button className="button">Save</button>
      </form>
    );
  };
};

export default CategoriesForm;
