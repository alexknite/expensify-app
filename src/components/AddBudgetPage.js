import React from 'react';
import { connect } from 'react-redux';
import BudgetForm from './BudgetForm';
import { startAddBudget } from '../actions/budgets';
import { selectBudget } from '../actions/budget';

export class AddBudgetPage extends React.Component {
  onSubmit = (budget) => {
    this.props.startAddBudget(budget);
    this.props.selectBudget({});
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Budget</h1>
          </div>
        </div>
        <div className="content-container">
          <BudgetForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddBudget: (budget) => dispatch(startAddBudget(budget)),
  selectBudget: (budget) => dispatch(startAddBudget(budget))
});

export default connect(undefined, mapDispatchToProps)(AddBudgetPage);
