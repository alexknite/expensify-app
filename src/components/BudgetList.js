import React from 'react';
import { connect } from 'react-redux';
import BudgetListItem from './BudgetListItem';
import selectBudgets from '../selectors/budgets';

export const BudgetList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Budgets</div>
      <div className="show-for-desktop">Budget</div>
    </div>
    <div className="list-body">
      {
        props.budgets.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No budgets available</span>
          </div>
        ) : (
          props.budgets.map((budget) => {
            return <BudgetListItem key={budget.id} {...budget} />
          })
        )
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    budgets: selectBudgets(state.budgets, state.budgetsFilters)
  };
};

export default connect(mapStateToProps)(BudgetList);
