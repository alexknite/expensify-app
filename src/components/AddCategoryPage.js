import React from 'react';
import { connect } from 'react-redux';
import CategoriesForm from './CategoriesForm';
import { startAddCategory } from '../actions/categories';

export class AddCategoryPage extends React.Component {
  onSubmit = (category) => {
    this.props.startAddCategory(category);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Category</h1>
          </div>
        </div>
        <div className="content-container">
          <CategoriesForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddCategory: (category) => dispatch(startAddCategory(category)),
});

export default connect(undefined, mapDispatchToProps)(AddCategoryPage);
