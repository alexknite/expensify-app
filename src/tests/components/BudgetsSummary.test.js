import React from 'react';
import { shallow } from 'enzyme';
import { BudgetsSummary } from '../../components/BudgetsSummary';

test('should correctly render BudgetsSummary with one budget', () => {
  const wrapper = shallow(
    <BudgetsSummary
      moneySpent={50}
      moneyAlotted={100}
      moneyLeft={50}
      category={'Food'}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
