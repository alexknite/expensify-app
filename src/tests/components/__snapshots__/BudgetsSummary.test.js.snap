// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render BudgetsSummary with one budget 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <div>
      <h1
        className="page-header__title"
      >
        You have spent 
        <span>
          $0.50
        </span>
         out of 
        <span>
          $1.00
        </span>
         for 
        <span>
          Food
        </span>
      </h1>
      <h1
        className="page-header__title"
      >
        You may spend 
        <span>
          $0.50
        </span>
         more
      </h1>
    </div>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create/budget"
      >
        Add Budget
      </Link>
    </div>
  </div>
</div>
`;
